File: apps/quickink/ios/QuickInk/Stories/StoriesShelfScreen.swift
/*
 * StoriesShelfScreen.swift
 *
 * The Stories tab — §7.1 of design/stories-mockup-v3.html.
 *
 *   ┌─────────────────────────────────────────┐
 *   │  Stories                                │
 *   │  ⌕  Search your stories                 │
 *   │                                         │
 *   │  ┌─ SUGGESTED · TODAY ────────────┐    │
 *   │  │ ▓▓▓ ▓ ▓                         │    │
 *   │  │ Tokyo, May 4–7                  │    │
 *   │  │ 12 photos and 3 receipts…       │    │
 *   │  │ Not interested      Open prev → │    │
 *   │  └─────────────────────────────────┘    │
 *   │                                         │
 *   │  Your stories                           │
 *   │  ┌────────────────────────────────┐    │
 *   │  │ ▓▓│ Mira's first month       › │    │  ← story card
 *   │  │   │ 14 items · Apr 2026        │    │
 *   │  │   │ [Public link]              │    │
 *   │  └────────────────────────────────┘    │
 *   │                                  ╭───╮ │
 *   │                                  │ + │ │  ← FAB
 *   │                                  ╰───╯ │
 *   └─────────────────────────────────────────┘
 *
 * Phase 1 surface — reads from `StoriesShelfViewModel`. Suggestion
 * hero card is shown only when a real suggestion exists (Phase 5
 * fills the cache); when the shelf is empty AND there is no
 * suggestion we render a calm placeholder hero introducing the
 * feature. Creation flow + editor land in Phase 2.
 *
 * Mirror of Android `StoriesShelfScreen.kt`.
 */

import SwiftUI

struct StoriesShelfScreen: View {

    let userId: String
    /// Phase 2 navigation hook — opens the story in the editor.
    /// Called both when the user taps a shelf card (passes its id)
    /// AND when the "+" FAB lands after creating a fresh draft.
    var onOpenStory: (String) -> Void = { _ in }
    /// Phase 5 — push the suggestion preview screen.
    var onOpenSuggestionPreview: (String) -> Void = { _ in }

    @StateObject private var vm: StoriesShelfViewModel
    @State private var searchDraft: String = ""

    init(
        userId: String,
        onOpenStory: @escaping (String) -> Void = { _ in },
        onOpenSuggestionPreview: @escaping (String) -> Void = { _ in }
    ) {
        self.userId = userId
        self.onOpenStory = onOpenStory
        self.onOpenSuggestionPreview = onOpenSuggestionPreview
        _vm = StateObject(wrappedValue: StoriesShelfViewModel(userId: userId))
    }

    var body: some View {
        ZStack(alignment: .bottomTrailing) {
            QuickInkColors.bg.ignoresSafeArea()

            ScrollView(showsIndicators: false) {
                VStack(alignment: .leading, spacing: QuickInkSpacing.s4) {
                    Text("Stories")
                        .font(QuickInkText.display)
                        .foregroundStyle(QuickInkColors.ink)
                        .padding(.top, QuickInkSpacing.s2)

                    searchBar

                    if let suggestion = vm.suggestion {
                        suggestionHero(suggestion: suggestion)
                    } else if vm.rows.isEmpty {
                        emptyStateHero
                    }

                    if !vm.rows.isEmpty {
                        sectionHeader("Your stories")
                        VStack(spacing: QuickInkSpacing.s2 + 2) {
                            ForEach(vm.rows, id: \.story.id) { row in
                                Button(action: { onOpenStory(row.story.id) }) {
                                    StoryShelfCard(row: row)
                                }
                                .buttonStyle(.plain)
                            }
                        }
                    }

                    Color.clear.frame(height: QuickInkBottomNavReservedHeight)
                }
                .padding(.horizontal, QuickInkSpacing.s4)
            }

            fab
        }
        .task {
            vm.start()
        }
    }

    // MARK: - Subviews

    private var searchBar: some View {
        HStack(spacing: QuickInkSpacing.s2) {
            Image(systemName: "magnifyingglass")
                .font(.system(size: 15))
                .foregroundStyle(QuickInkColors.inkSoft)
            TextField("Search your stories", text: $searchDraft)
                .font(.system(size: 14))
                .foregroundStyle(QuickInkColors.ink)
                .textInputAutocapitalization(.never)
                .disableAutocorrection(true)
        }
        .padding(.horizontal, QuickInkSpacing.s3)
        .padding(.vertical, QuickInkSpacing.s3 - 2)
        .background(
            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                .fill(QuickInkColors.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.md, style: .continuous)
                .strokeBorder(QuickInkColors.border, lineWidth: 1)
        )
    }

    private func suggestionHero(suggestion: StorySuggestion) -> some View {
        VStack(alignment: .leading, spacing: QuickInkSpacing.s2) {
            Text("SUGGESTED · TODAY")
                .font(.system(size: 10, weight: .medium))
                .tracking(1.5)
                .foregroundStyle(QuickInkColors.accentDeep)

            previewStrip

            Text(suggestion.reason)
                .font(QuickInkText.editorial)
                .foregroundStyle(QuickInkColors.ink)
                .lineLimit(2)

            HStack {
                Button("Not interested") {
                    vm.dismissSuggestion()
                }
                .font(QuickInkText.bodyItalic)
                .foregroundStyle(QuickInkColors.muted)
                .buttonStyle(.plain)
                Spacer()
                Button("Open preview →") {
                    onOpenSuggestionPreview(suggestion.id)
                }
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(QuickInkColors.accent)
                .buttonStyle(.plain)
            }
            .padding(.top, QuickInkSpacing.s1)
            .overlay(alignment: .top) {
                Rectangle()
                    .fill(QuickInkColors.border)
                    .frame(height: 0.5)
            }
        }
        .padding(QuickInkSpacing.s3)
        .background(
            RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous)
                .fill(QuickInkColors.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous)
                .strokeBorder(QuickInkColors.border, lineWidth: 1)
        )
    }

    /// Calm placeholder shown when the shelf has no stories AND the
    /// suggestion engine hasn't produced anything yet. Introduces the
    /// feature without faking a suggestion.
    private var emptyStateHero: some View {
        VStack(alignment: .leading, spacing: QuickInkSpacing.s2) {
            Text("YOUR STORIES")
                .font(.system(size: 10, weight: .medium))
                .tracking(1.5)
                .foregroundStyle(QuickInkColors.accentDeep)

            Text("Curate a moment.")
                .font(QuickInkText.editorial)
                .foregroundStyle(QuickInkColors.ink)

            Text("Pick a few scans, photos, or notes and assemble them into a story you can share.")
                .font(QuickInkText.bodyItalic)
                .foregroundStyle(QuickInkColors.inkSoft)
                .fixedSize(horizontal: false, vertical: true)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(QuickInkSpacing.s4)
        .background(
            RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous)
                .fill(QuickInkColors.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.lg, style: .continuous)
                .strokeBorder(QuickInkColors.border, lineWidth: 1)
        )
    }

    /// Mockup §7.1 preview-strip — a 2:1:1 grid of paper / paper / coral
    /// thumbnails standing in for "the first three items of the
    /// suggested cluster". Phase 5 will render real captures here.
    private var previewStrip: some View {
        HStack(spacing: 4) {
            RoundedRectangle(cornerRadius: 4)
                .fill(QuickInkColors.paper1)
                .frame(maxWidth: .infinity)
            RoundedRectangle(cornerRadius: 4)
                .fill(QuickInkColors.paper3)
                .frame(maxWidth: .infinity)
            RoundedRectangle(cornerRadius: 4)
                .fill(QuickInkColors.accent.opacity(0.55))
                .frame(maxWidth: .infinity)
        }
        .frame(height: 70)
        .clipShape(RoundedRectangle(cornerRadius: 8, style: .continuous))
    }

    private func sectionHeader(_ text: String) -> some View {
        Text(text)
            .font(QuickInkText.editorial)
            .foregroundStyle(QuickInkColors.ink)
            .padding(.top, QuickInkSpacing.s1)
    }

    /// Lifted coral disc, bottom-right. Phase 1 stubs the create
    /// action; the real flow comes in Phase 2.
    private var fab: some View {
        Button(action: {
            Task {
                if let id = await vm.createDraft() {
                    onOpenStory(id)
                }
            }
        }) {
            ZStack {
                Circle()
                    .fill(LinearGradient(
                        colors: [QuickInkColors.accent, QuickInkColors.accentDeep],
                        startPoint: .top,
                        endPoint: .bottom
                    ))
                    .frame(width: 56, height: 56)
                    .shadow(color: QuickInkColors.accent.opacity(0.35), radius: 12, x: 0, y: 6)
                Image(systemName: "plus")
                    .font(.system(size: 24, weight: .semibold))
                    .foregroundStyle(QuickInkColors.textOnAccent)
            }
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Create a new story")
        // Sit clear of the bottom-nav FAB centre while still reading
        // as "on the same level" — matches the mockup's bottom-right
        // anchor (~24 from edges, lifted above the nav).
        .padding(.trailing, QuickInkSpacing.s4)
        .padding(.bottom, QuickInkBottomNavReservedHeight - 32)
    }
}

// MARK: - Story shelf card

private struct StoryShelfCard: View {
    let row: StoryShelfRow

    var body: some View {
        HStack(spacing: QuickInkSpacing.s3) {
            cover
            VStack(alignment: .leading, spacing: 2) {
                Text(row.story.title)
                    .font(QuickInkText.editorial)
                    .foregroundStyle(QuickInkColors.ink)
                    .lineLimit(1)
                Text(metaLine)
                    .font(.system(size: 11))
                    .foregroundStyle(QuickInkColors.inkSoft)
                sharePill
            }
            Spacer(minLength: 0)
        }
        .padding(QuickInkSpacing.s2 + 2)
        .background(
            RoundedRectangle(cornerRadius: QuickInkRadius.md + 2, style: .continuous)
                .fill(QuickInkColors.surface)
        )
        .overlay(
            RoundedRectangle(cornerRadius: QuickInkRadius.md + 2, style: .continuous)
                .strokeBorder(QuickInkColors.border, lineWidth: 1)
        )
    }

    // MARK: cover

    private var cover: some View {
        RoundedRectangle(cornerRadius: 10, style: .continuous)
            .fill(coverFill)
            .frame(width: 56, height: 56)
            .overlay(alignment: .bottomLeading) {
                Text(coverCaption)
                    .font(QuickInkFont.serif(7.5, weight: .medium))
                    .foregroundStyle(coverCaptionTint)
                    .padding(.horizontal, 5)
                    .padding(.bottom, 4)
            }
    }

    /// Phase 1 cover palette — Story.coverStyle picks the colour
    /// (photo + no cover_item_id → paper-warm; gradient → coral wash;
    /// typographic → tan paper). Phase 3's dominant-colour cover will
    /// supersede this with a real hero colour.
    private var coverFill: Color {
        switch row.story.coverStyle {
        case Story.CoverStyle.gradient.rawValue:    return QuickInkColors.accent.opacity(0.7)
        case Story.CoverStyle.typographic.rawValue: return QuickInkColors.paper3
        default:                                    return QuickInkColors.paper1
        }
    }

    private var coverCaptionTint: Color {
        row.story.coverStyle == Story.CoverStyle.gradient.rawValue
            ? QuickInkColors.textOnAccent
            : QuickInkColors.ink
    }

    /// Two-or-three-letter handle pulled from the title (first word).
    /// Mirrors the mockup's "Mira" / "Lisbon" / "Reno" cover labels.
    private var coverCaption: String {
        let first = row.story.title.split(separator: " ").first.map(String.init) ?? row.story.title
        return String(first.prefix(8))
    }

    // MARK: meta + pill

    /// "14 items · Apr 2026" — item count + month-year of the latest
    /// item, or "no items yet" when the shelf row hasn't accumulated
    /// any children yet.
    private var metaLine: String {
        let countPart: String = {
            switch row.itemCount {
            case 0:  return "no items yet"
            case 1:  return "1 item"
            default: return "\(row.itemCount) items"
            }
        }()
        guard let isoLatest = row.latestItemAt,
              let monthYear = Self.monthYear(fromIso: isoLatest)
        else { return countPart }
        return "\(countPart) · \(monthYear)"
    }

    @ViewBuilder
    private var sharePill: some View {
        switch row.story.shareMode {
        case Story.ShareMode.publicLink.rawValue:
            pillView(text: "Public link", tone: .live)
        case Story.ShareMode.exported.rawValue,
             Story.ShareMode.inApp.rawValue:
            pillView(text: "Shared", tone: .live)
        case Story.ShareMode.private.rawValue
            where row.story.status == Story.Status.draft.rawValue:
            pillView(text: "Draft", tone: .neutral)
        default:
            EmptyView()
        }
    }

    private enum PillTone { case live, neutral }

    private func pillView(text: String, tone: PillTone) -> some View {
        let bg = tone == .live ? QuickInkColors.accentSoft : QuickInkColors.borderSoft
        let fg = tone == .live ? QuickInkColors.accent     : QuickInkColors.inkSoft
        return Text(text)
            .font(.system(size: 10, weight: .medium))
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(bg, in: Capsule())
            .foregroundStyle(fg)
            .padding(.top, 4)
    }

    // MARK: - Helpers

    private static let isoParser: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let monthYearFmt: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM yyyy"
        return f
    }()

    private static func monthYear(fromIso iso: String) -> String? {
        // The schema stores fractional-seconds ISO-8601; older
        // GRDB defaults (and Drive payloads from earlier clients)
        // can land without the fraction. Parse the strict form
        // first, fall back to the non-fractional formatter so
        // either shape renders.
        if let d = isoParser.date(from: iso) { return monthYearFmt.string(from: d) }
        let fallback = ISO8601DateFormatter()
        fallback.formatOptions = [.withInternetDateTime]
        if let d = fallback.date(from: iso) { return monthYearFmt.string(from: d) }
        return nil
    }
}

#if DEBUG
struct StoriesShelfScreen_Previews: PreviewProvider {
    static var previews: some View {
        StoriesShelfScreen(userId: "preview-user")
    }
}
#endif

File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoriesShelfScreen.kt
/*
 * StoriesShelfScreen.kt
 *
 * The Stories tab — §7.1 of design/stories-mockup-v3.html. Mirror of
 * iOS `StoriesShelfScreen.swift`; see that file's header for the
 * ASCII layout.
 *
 * Phase 1 surface: reads from [StoriesShelfViewModel], renders the
 * page title, search bar (non-functional placeholder for Phase 1),
 * suggestion hero card (only when a suggestion exists — Phase 5
 * fills the cache), empty-state hero (when both rows + suggestion
 * are absent), the "Your stories" list, and a coral "+" FAB whose
 * tap target is stubbed (Phase 2 wires the creation flow).
 */

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.quickink.mobile.features.stories

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import kotlinx.coroutines.launch
import app.quickink.mobile.data.story.StoryEntity
import app.quickink.mobile.data.story.StoryShelfRow
import app.quickink.mobile.features.nav.QuickInkBottomNavReservedHeight
import app.quickink.mobile.ui.theme.LocalQuickInkColors
import app.quickink.mobile.ui.theme.LocalQuickInkTypography
import app.quickink.mobile.ui.theme.QuickInkRadius
import app.quickink.mobile.ui.theme.QuickInkSpacing
import java.time.OffsetDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

@Composable
fun StoriesShelfScreen(
    userId: String,
    /** Phase 2 navigation hook — opens the story in the editor.
     *  Called both when the user taps a shelf card (passes its id)
     *  AND when the "+" FAB lands after creating a fresh draft. */
    onOpenStory: (String) -> Unit = { _ -> },
    /** Phase 5 — push the suggestion preview screen. */
    onOpenSuggestionPreview: (String) -> Unit = { _ -> },
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current

    val vm: StoriesShelfViewModel = viewModel(
        factory = StoriesShelfViewModel.factory(userId),
    )
    val rows by vm.rows.collectAsState()
    val suggestion by vm.suggestion.collectAsState()

    var searchDraft by remember { mutableStateOf("") }
    val scope = androidx.compose.runtime.rememberCoroutineScope()

    Box(
        modifier = Modifier
            .fillMaxSize()
            .background(colors.bg),
    ) {
        LazyColumn(
            modifier = Modifier.fillMaxSize(),
            contentPadding = PaddingValues(
                start  = QuickInkSpacing.s4,
                end    = QuickInkSpacing.s4,
                top    = QuickInkSpacing.s2,
                bottom = QuickInkBottomNavReservedHeight,
            ),
            verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s4),
        ) {
            item("title") {
                Text(
                    text  = "Stories",
                    style = type.display,
                    color = colors.ink,
                )
            }

            item("search") {
                SearchBar(
                    value          = searchDraft,
                    onValueChange  = { searchDraft = it },
                )
            }

            if (suggestion != null) {
                item("hero") {
                    SuggestionHero(
                        reason     = suggestion!!.reason,
                        onDismiss  = { vm.dismissSuggestion() },
                        onPreview  = { onOpenSuggestionPreview(suggestion!!.id) },
                    )
                }
            } else if (rows.isEmpty()) {
                item("hero-empty") { EmptyStateHero() }
            }

            if (rows.isNotEmpty()) {
                item("section-header") {
                    Text(
                        text  = "Your stories",
                        style = type.editorial,
                        color = colors.ink,
                        modifier = Modifier.padding(top = QuickInkSpacing.s1),
                    )
                }
                items(rows, key = { it.story.id }) { row ->
                    Box(
                        modifier = Modifier.clickable { onOpenStory(row.story.id) },
                    ) {
                        StoryShelfCard(row = row)
                    }
                }
            }
        }

        Fab(
            onClick = {
                scope.launch {
                    val id = vm.createDraft()
                    if (id != null) onOpenStory(id)
                }
            },
            modifier = Modifier
                .align(Alignment.BottomEnd)
                .padding(end = QuickInkSpacing.s4)
                .padding(bottom = QuickInkBottomNavReservedHeight - 32.dp),
        )
    }
}

// MARK: - Search bar

@Composable
private fun SearchBar(value: String, onValueChange: (String) -> Unit) {
    val colors = LocalQuickInkColors.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(QuickInkRadius.md))
            .background(colors.surface)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.md))
            .padding(horizontal = QuickInkSpacing.s3, vertical = QuickInkSpacing.s2 + 2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Icon(
            imageVector        = Icons.Filled.Search,
            contentDescription = null,
            tint               = colors.inkSoft,
            modifier           = Modifier.size(18.dp),
        )
        Spacer(modifier = Modifier.width(QuickInkSpacing.s2))
        Box(modifier = Modifier.fillMaxWidth()) {
            if (value.isEmpty()) {
                Text(
                    text  = "Search your stories",
                    color = colors.inkSoft,
                    fontSize = 14.sp,
                )
            }
            BasicTextField(
                value            = value,
                onValueChange    = onValueChange,
                singleLine       = true,
                textStyle        = androidx.compose.ui.text.TextStyle(
                    color    = colors.ink,
                    fontSize = 14.sp,
                ),
                cursorBrush      = androidx.compose.ui.graphics.SolidColor(colors.accent),
                keyboardOptions  = KeyboardOptions.Default,
                modifier         = Modifier.fillMaxWidth(),
            )
        }
    }
}

// MARK: - Suggestion hero (Phase 5 fills this; v3 ships with it always
// off and the empty-state hero in its place)

@Composable
private fun SuggestionHero(
    reason: String,
    onDismiss: () -> Unit,
    onPreview: () -> Unit,
) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(QuickInkRadius.lg))
            .background(colors.surface)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.lg))
            .padding(QuickInkSpacing.s3),
        verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s2),
    ) {
        Text(
            text       = "SUGGESTED · TODAY",
            color      = colors.accentDeep,
            fontSize   = 10.sp,
            fontWeight = FontWeight.Medium,
            letterSpacing = 1.5.sp,
        )
        PreviewStrip()
        Text(
            text       = reason,
            style      = type.editorial,
            color      = colors.ink,
            maxLines   = 2,
            overflow   = TextOverflow.Ellipsis,
        )
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(top = QuickInkSpacing.s1),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text  = "Not interested",
                style = type.bodyItalic,
                color = colors.muted,
                fontSize = 11.sp,
                modifier = Modifier.clickable(onClick = onDismiss),
            )
            Spacer(modifier = Modifier.weight(1f))
            Text(
                text       = "Open preview →",
                color      = colors.accent,
                fontSize   = 12.sp,
                fontWeight = FontWeight.Medium,
                modifier = Modifier.clickable(onClick = onPreview),
            )
        }
    }
}

@Composable
private fun EmptyStateHero() {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(QuickInkRadius.lg))
            .background(colors.surface)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.lg))
            .padding(QuickInkSpacing.s4),
        verticalArrangement = Arrangement.spacedBy(QuickInkSpacing.s2),
    ) {
        Text(
            text       = "YOUR STORIES",
            color      = colors.accentDeep,
            fontSize   = 10.sp,
            fontWeight = FontWeight.Medium,
            letterSpacing = 1.5.sp,
        )
        Text(
            text  = "Curate a moment.",
            style = type.editorial,
            color = colors.ink,
        )
        Text(
            text  = "Pick a few scans, photos, or notes and assemble them into a story you can share.",
            style = type.bodyItalic,
            color = colors.inkSoft,
        )
    }
}

@Composable
private fun PreviewStrip() {
    val colors = LocalQuickInkColors.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(70.dp)
            .clip(RoundedCornerShape(8.dp)),
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Box(
            modifier = Modifier
                .weight(2f)
                .fillMaxSize()
                .clip(RoundedCornerShape(4.dp))
                .background(colors.paper1),
        )
        Box(
            modifier = Modifier
                .weight(1f)
                .fillMaxSize()
                .clip(RoundedCornerShape(4.dp))
                .background(colors.paper3),
        )
        Box(
            modifier = Modifier
                .weight(1f)
                .fillMaxSize()
                .clip(RoundedCornerShape(4.dp))
                .background(colors.accent.copy(alpha = 0.55f)),
        )
    }
}

// MARK: - Story shelf card

@Composable
private fun StoryShelfCard(row: StoryShelfRow) {
    val colors = LocalQuickInkColors.current
    val type   = LocalQuickInkTypography.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(QuickInkRadius.md + 2.dp))
            .background(colors.surface)
            .border(1.dp, colors.border, RoundedCornerShape(QuickInkRadius.md + 2.dp))
            .padding(QuickInkSpacing.s2 + 2.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(QuickInkSpacing.s3),
    ) {
        Cover(story = row.story)
        Column(modifier = Modifier.weight(1f)) {
            Text(
                text       = row.story.title,
                style      = type.editorial,
                color      = colors.ink,
                maxLines   = 1,
                overflow   = TextOverflow.Ellipsis,
            )
            Text(
                text  = metaLine(row),
                color = colors.inkSoft,
                fontSize = 11.sp,
            )
            SharePill(story = row.story)
        }
    }
}

@Composable
private fun Cover(story: StoryEntity) {
    val colors = LocalQuickInkColors.current
    val fill = when (story.coverStyle) {
        StoryEntity.CoverStyle.GRADIENT.raw    -> colors.accent.copy(alpha = 0.7f)
        StoryEntity.CoverStyle.TYPOGRAPHIC.raw -> colors.paper3
        else                                   -> colors.paper1
    }
    val captionTint = if (story.coverStyle == StoryEntity.CoverStyle.GRADIENT.raw)
        colors.textOnAccent else colors.ink
    Box(
        modifier = Modifier
            .size(56.dp)
            .clip(RoundedCornerShape(10.dp))
            .background(fill),
        contentAlignment = Alignment.BottomStart,
    ) {
        Text(
            text       = coverCaption(story.title),
            color      = captionTint,
            fontSize   = 7.5.sp,
            fontWeight = FontWeight.Medium,
            modifier   = Modifier.padding(horizontal = 5.dp, vertical = 4.dp),
        )
    }
}

@Composable
private fun SharePill(story: StoryEntity) {
    val colors = LocalQuickInkColors.current
    val (text, bg, fg) = when {
        story.shareMode == StoryEntity.ShareMode.PUBLIC_LINK.raw ->
            Triple("Public link", colors.accentSoft, colors.accent)
        story.shareMode == StoryEntity.ShareMode.EXPORTED.raw ||
            story.shareMode == StoryEntity.ShareMode.IN_APP.raw ->
            Triple("Shared", colors.accentSoft, colors.accent)
        story.status == StoryEntity.Status.DRAFT.raw ->
            Triple("Draft", colors.borderSoft, colors.inkSoft)
        else -> return
    }
    Box(
        modifier = Modifier
            .padding(top = 4.dp)
            .clip(RoundedCornerShape(QuickInkRadius.pill))
            .background(bg)
            .padding(horizontal = 8.dp, vertical = 3.dp),
    ) {
        Text(
            text       = text,
            color      = fg,
            fontSize   = 10.sp,
            fontWeight = FontWeight.Medium,
        )
    }
}

// MARK: - FAB

@Composable
private fun Fab(onClick: () -> Unit, modifier: Modifier = Modifier) {
    val colors = LocalQuickInkColors.current
    val gradient = Brush.verticalGradient(
        colors = listOf(colors.accent, colors.accentDeep),
    )
    val interactionSource = remember { MutableInteractionSource() }
    Box(
        modifier = modifier
            .size(56.dp)
            .shadow(elevation = 12.dp, shape = CircleShape, clip = false)
            .clip(CircleShape)
            .background(gradient)
            .clickable(
                interactionSource = interactionSource,
                indication        = null,
                onClick           = onClick,
            ),
        contentAlignment = Alignment.Center,
    ) {
        Icon(
            imageVector        = Icons.Filled.Add,
            contentDescription = "Create a new story",
            tint               = colors.textOnAccent,
            modifier           = Modifier.size(24.dp),
        )
    }
}

// MARK: - Helpers

private fun coverCaption(title: String): String {
    val first = title.split(' ').firstOrNull()?.takeIf { it.isNotEmpty() } ?: title
    return first.take(8)
}

private fun metaLine(row: StoryShelfRow): String {
    val countPart = when (row.itemCount) {
        0    -> "no items yet"
        1    -> "1 item"
        else -> "${row.itemCount} items"
    }
    val monthYear = row.latestItemAt?.let { parseMonthYear(it) }
    return if (monthYear == null) countPart else "$countPart · $monthYear"
}

private val MONTH_YEAR_FMT = DateTimeFormatter.ofPattern("MMM yyyy", Locale.ENGLISH)

/**
 * Parse the schema's ISO-8601 timestamps into "Apr 2026". The
 * canonical writer (`IsoClock.nowIso`) emits fractional seconds; we
 * accept either shape via `OffsetDateTime.parse`, which handles
 * both.
 */
private fun parseMonthYear(iso: String): String? = runCatching {
    OffsetDateTime.parse(iso).format(MONTH_YEAR_FMT)
}.getOrNull()
